File: 56. Merge Intervals.cpp
/**
 * Definition for an interval.
 * struct Interval {
 *     int start;
 *     int end;
 *     Interval() : start(0), end(0) {}
 *     Interval(int s, int e) : start(s), end(e) {}
 * };
 */
class Solution {
public:
    static bool myCompare(Interval& a, Interval &b) {
        return a.start < b.start || (a.start == b.start && a.end < b.end);
    }
    vector<Interval> merge(vector<Interval>& intervals) {
        if (intervals.size() < 2) return intervals; 
        sort(intervals.begin(), intervals.end(), myCompare);
        vector<Interval> res;
        res.push_back(intervals[0]);
        int i = 1;
        while (i < intervals.size()) {
            if (intervals[i].start <= res[res.size() - 1].end) {
                res[res.size() - 1].end = max(res[res.size() - 1].end, intervals[i].end);
            } else {
                res.push_back(intervals[i]);
            }
            i++;
        }
        return res;
    }
};

File: java/56. Merge Intervals.java
/**
 * Definition for an interval.
 * public class Interval {
 *     int start;
 *     int end;
 *     Interval() { start = 0; end = 0; }
 *     Interval(int s, int e) { start = s; end = e; }
 * }
 */
// O(n)
class Solution {
    public List<Interval> merge(List<Interval> intervals) {
        List<Interval> res = new ArrayList<>();
        intervals.sort(new Comparator<Interval>() {
            public int compare(Interval a, Interval b) {
                return a.start - b.start;
            }
        });
        int i = 0;
        while (i < intervals.size()) {
            int j = i + 1, right = intervals.get(i).end;
            while (j < intervals.size() && intervals.get(j).start <= right) {
                right = Math.max(right, intervals.get(j).end);
                ++j;
            }
            res.add(new Interval(intervals.get(i).start, right));
            i = j;
        }
        return res;
    }
}
